Remove compass and replace with Bourbon

maxtedford opened the pull request:
This work addresses issue #1037 that was raised by my colleague, by tearing out sassy-buttons and compass and copying over or replacing their related mixins and CSS accordingly. To assist with rebuilding some of the CSS, we enlisted the help of [Bourbon](http://bourbon.io/docs). We did our best to integrate it in a non-intrusive way that should allow for you and the rest of the team to easily switch to Bootstrap or Foundation should you decide to do so for your next major release. This will also allow us to make use of a faster Sass pre-processor like [SassC::Rails](https://github.com/sass/sassc-rails).

In addition, we noticed that this work may also have a positive residual impact on issue #939, which makes reference to compass-rails needing to be updated to support Rails 5.

We look forward to your feedback. Thanks in advance!

cc: @metaskills

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://customink-removecompass.alchemycms-alchemy-cms-42e8f4fc945384e0f6b5.ttrcloud.com/admin' target='_blank'>Open stage</a> or <a href='http://customink-removecompass.alchemycms-alchemy-cms-42e8f4fc945384e0f6b5.ttrcloud.com/admin/teatro.log' target='_blank'>view logs</a>.

tvdeyen commented:
First of all: Thank you for your contribution and effort in this. Very much appreciated.

I have some thoughts on this:

#### 1. Why did you removed `sassy-buttons` without replacement?

You said you:
> copying over or replacing their related mixins and CSS accordingly. To assist with rebuilding some of the CSS, we enlisted the help of Bourbon

I pulled your branch and saw the buttons and selects all look unstyled, like this:

![unstyled-buttons](https://cloud.githubusercontent.com/assets/42868/15944294/3af36028-2e8d-11e6-95a4-bd57378336bb.jpg)

It seems that you just removed `sassy-buttons` without replacing the mixin with the style we where using. Is there any particular reason?

To clarify: We want to keep the current look of Alchemy. Redesign is planned for next mayor version (4.0). Until than we need to keep the same look.

#### 2. Why don't you use `bourbon` mixins?

Bourbon has lots of useful mixins ([`clearfix`](http://bourbon.io/docs/#clearfix) or the handy [`border-radius` shortcuts](http://bourbon.io/docs/#border-radius)). If we add a dependency we should leverage it's strengths. Just having a dependency for nearly no use does not make sense to me.

#### 3. You removed mixins without replacement

Some mixins from `compass` has no Bourbon counterpart, like the deprecated `inline-block` mixin. You replaced them with own css rules. This introduces a lot of additional css rules. Why not copy the mixins over, so we can keep on using it?

**Don't get me wrong**, I *really* appreciate the work, but unfortunately I can't merge this in until we:

1. Keep the default button and select styles by copying over the `sassy-buttons` styles into our own button mixin. *Alternatively* we could use the [simple button mixin from bourbon](http://bourbon.io/docs/#buttons), although it's deprecated. (If we plan to use Bootstrap 4 over Bourbon 5 anyway it's ok, as long as we keep the change as small as possible for now. And if we decide to use Bourbon 5 with Bitters we can migrate later on.)
2. Using as much Bourbon mixins and helpers as we can to reduce the duplication of rules *or* create own mixins and don't use Bourbon at all (especially if we want to replace it with Bootstrap 4 anyway?)
3. Create own mixins for Compass mixins Bourbon does not provide.

My apologies that we haven't decided to go either with Bootstrap, Bourbon/Bitters or Foundation now. 

I think it's ok to go with Bourbon for now. 

### Thanks

I would also like to hear the cents of @robinboening, @wdspkr & @mamhoff WDYT?

tvdeyen commented on alchemy_cms.gemspec:
```diff
@@ -39,7 +39,6 @@ Gem::Specification.new do |gem|
   gem.add_runtime_dependency 'request_store',                    ['~> 1.2']
   gem.add_runtime_dependency 'responders',                       ['~> 2.0']
   gem.add_runtime_dependency 'sass-rails',                       ['~> 5.0']
```
We also should remove the `sass-rails` dependency, so one can more easily replace it with `sassc-rails` in their projects.

tvdeyen commented on lib/alchemy/engine.rb:
```diff
@@ -16,7 +16,6 @@
 require 'request_store'
 require 'responders'
 require 'sass-rails'
```
We shouldn't depend on `sass-rails`

mamhoff commented:
I think it's alright to drop supporting old browsers and not using mixins for `border-radius` and the likes as those are supported by all major browsers right now. 

I agree with @tvdeyen on keeping the button styles. They do work well with the current design, and we don't want to move too iteratively on that one (especially as we don't have a roadmap that tells us where to go yet). 

Otherwise, this is wonderful work, thank you! 🎊

tvdeyen commented:
Sure. Only use mixins that are necessary. Bourbon has pretty good deprecation notices for mixins that are not needed any more, because the browser support is good enough.

maxtedford commented:
Thanks so much for the fantastic feedback. We're very happy to help! We've made a couple changes based on your comments/questions. As I explain, allow me to answer the questions you posited:

## **1. Why we didn't fully replace sassy-buttons** 

We didn't fully understand up front that your aim was to keep all the styles as close to exactly the same as they were with sassy-buttons, so we figured the styles were close enough to suffice. Now that we have a clearer understanding that you'd like to avoid even minor iterative style changes until the next major release, we've made use of Bourbon's `button` mixin to conform the style nearly exactly to what you created before the switch to bourbon. As you mentioned, that should allow you to easily integrate with Bourbon/Bitters should you decide to go that route, or make use of the Bootstrap or Foundation equivalents. Screenshots below for you:

**compass/sassy-buttons:**
![before_bourbon](https://cloud.githubusercontent.com/assets/9435299/15976475/ff729868-2f20-11e6-8b52-93465be7007e.png)

**bourbon:**
![after_bourbon](https://cloud.githubusercontent.com/assets/9435299/15976481/0632b282-2f21-11e6-825f-4a237449b322.png)

## **2. Why we don't use Bourbon mixins**

We didn't use a ton of Bourbon mixins for pretty much the same reason as above. We heard you say that you haven't decided on a style roadmap at this point, so we didn't want to overly-commit you to  bourbon-specific mixins before you were ready to finalize your framework of choice. With that said, your point is well taken that it doesn't make a ton of sense to depend on the framework, but not make use of its strengths, so I went through and reverted some a few of the mixins I removed originally, replacing them with some of their bourbon equivalents. If you're interested in making even more use of bourbon, that's certainly possible, but figured I'd defer to you before going too crazy. There are also quite a few mixins used by Compass that are no longer necessary since current major browsers support their direct CSS equivalents, so I didn't revert those, choosing to rely on the pure css.

## **3. Why we didn't replace some of the mixins we removed**

Same as above for the the most part -- since al major browsers have support for things like `border-radius` and `box-shadow`, we didn't think it necessary to continue to use mixins for that purpose. It seems you agree for the most part, based on the comments above, so I don't think there's too much need to explain this one any further.

----

Additionally, you'll also notice that commit 03a293c introduces `sassc-rails` to the project. I benchmarked the pre-processing of `admin.scss` before including `sassc-rails`, then after including it, and saw the pre-processing time go from 5.4 seconds down to 2.0. This is a really nice win! Should make project much snappier 😃 🎉 

Hopefully this helps explain some of our thought process. The two new commits you'll see above hopefully address those concerns and you'll feel more comfortable proceeding with a merge.

cc @metaskills

metaskills commented:
@tvdeyen @maxtedford Agreed on removing both `sass-rails` and `sassc-rails` from the gem spec. Letting the [user decide seems to be the best choice](https://github.com/sass/sassc-rails#installing-alongside-a-gem-that-depends-on-sass-rails).

What is the best way to help users using Alchemy know they have to install some form of Sass Rails support? May I suggest doing something like this inside `lib/alchemy/engine.rb` where the removal of `sass-rails` occurred?

```ruby
begin
  require 'sassc-rails'
rescue LoadError, Gem::LoadError => e
  begin
    require 'sass-rails'
  rescue LoadError, Gem::LoadError => e
    raise LoadError, "Could not find the `sass-rails` or `sassc-rails` gem! Please add either to your project's Gemfile."
  end
end
```

tvdeyen commented:
@metaskills @maxtedford I like the idea of displaying a warning if no sass support has been found. 

But, isn't this a rare edge case? Every new rails app has sass support enabled. And even when it was removed, what happens when you start a rails app without sass support? Isn't the error from sprockets good enough? 

We should test this before adding more code for this (very unlikely) case.

metaskills commented:
> But, isn't this a rare edge case?

Not sure. I authored the `less-rails` gem and am quite surprised to see it has so much usage to date. Despite not liking Less.js myself at all. PostCSS seems to be rising in popularity and IMO, this is not an edge case and something we can easily guard against.

> Isn't the error from sprockets good enough?

Good point, but I think it helps give the user context on who needs Sass. Hence I think this project should do this at require time. I think this is both beneficial to your users and this gem by doing the require at the right time too.

tvdeyen commented:
Ok. Valid concerns. Let's do this. 

I would like to have this in a separate file we require from `engine.rb` then. For instance `sass_support.rb`. Also I hope there is some better way to check sass support than using costly requires and rescues (twice!) 

Any ideas?

metaskills commented:
Let's see what @maxtedford comes up with as I assume he is gonna drive this home early next week :)

Thanks so much y'all for letting us contribute this too. Cheers.

tvdeyen commented:
@metaskills @maxtedford great. Thanks for fixing the button styles. 

There is still some small issue, though:

![alchemy cms - seiten 2016-06-22 18-34-15](https://cloud.githubusercontent.com/assets/42868/16274940/1e769990-38a8-11e6-8ab7-7ccc1ca4bb3c.jpg)

Please use the `$default-border-color` variable on the buttons.

Any news on the require `sassc-rails` with fallback front?

I really like to have this in 3.5

Thanks again for all the work!

maxtedford commented:
@tvdeyen, sorry for the delay in finishing this up for you. I just pushed up three more commits.

The first one removes the `sassc-rails` dependency from the gemspec. Great catch on that, that was my mistake!

The second one restyles the `border-color` for both selects and buttons to the `$default-border-color` mixin you mentioned.

The third one implements the `sass-support.rb` file you requested. I have to admit that handling load error exceptions is not my forte, and after doing a little digging into potential alternatives, I ended up settling on something very similar to @metaskills's recommendation. I understand your concern about using costly requires and exceptions, but didn't uncover any cleaner alternatives, so let me know if you have any other thoughts! Additionally, I put the `require_relative 'sass-support'` line in `lib/alchemy/engine.rb` in the `#hacks` section, since that was this felt like to me. Please let me know if that should go somewhere else or you have an idea for organizational structure that I don't understand.

Thanks again for letting us work on this with you! 😄 We really appreciate what you've built here and we're very excited to contribute.

tvdeyen commented on app/assets/stylesheets/alchemy/selects.scss:
```diff
@@ -17,7 +17,7 @@ select {
       @include button($base-color: $white, $padding: 0 5px);
       display: block;
       font-weight: normal;
-      border-color: $button-border-color;
```
Oh, my bad. The buttons and selects should use `$button-border-color`, not `$default-border-color`, sorry.

Actually there are [lots of `button` variables](https://github.com/AlchemyCMS/alchemy_cms/blob/master/app/assets/stylesheets/alchemy/_variables.scss#L31-L36) we should use to style the buttons.

tvdeyen replied:
> Actually there are lots of button variables we should use to ~~style the buttons~~.

To override the bourbon buttons defaults.

maxtedford replied:
Cool, no problem. I can switch back super easily. Were there any specific button variables you had in mind to override the Bourbon defaults? I think everything looks preeeetty close right now, so I want to make sure I'm getting exactly what you're looking for so we can knock this out and get these things in 3.5 as you requested.

maxtedford replied:
> Were there any specific button variables you had in mind to override the Bourbon defaults?

I'm now realizing you linked to these, so nevermind. I'll see about incorporating some and pushing up another commit.

tvdeyen replied:
Great. Thanks. I apologize for all the mess, but UI is always the hardest ;)

maxtedford replied:
No worries! UI is _definitely_ the biggest pain in the butt with this stuff.

tvdeyen commented on lib/alchemy/engine.rb:
```diff
@@ -52,6 +51,7 @@
 
 # Require hacks
 require_relative './kaminari/scoped_pagination_url_helper'
+require_relative './sass-support'
```
I would not consider this a hack. Please just put it on line 47, thanks

maxtedford replied:
Fair enough. This felt a little off to me too as I was doing it, so I will move this line 47.

tvdeyen commented on lib/alchemy/sass-support.rb:
```diff
@@ -0,0 +1,11 @@
+module Alchemy
```
There is no need in scoping this into the `Alchemy` module, no?

maxtedford replied:
You're correct that there should be no need. I can remove that for you.

tvdeyen commented on lib/alchemy/sass-support.rb:
```diff
@@ -0,0 +1,11 @@
+    begin
+      require 'sass-rails'
+    rescue LoadError, Gem::LoadError => e
+      raise LoadError, "Could not find the `sass-rails` or `sassc-rails` gem! Please add one of them to your project's Gemfile."
```
Should we add " for AlchemyCMS" into this error message, so users can see more easily where this is coming from?

maxtedford replied:
Sure! That should help keep things explicit for the user. I can make that change no problem.

tvdeyen commented on lib/alchemy/sass-support.rb:
```diff
@@ -0,0 +1,11 @@
+  rescue LoadError, Gem::LoadError => e
+    begin
+      require 'sass-rails'
+    rescue LoadError, Gem::LoadError => e
```
Hound is in deed right. We don't need the `e` variable.

tvdeyen commented:
@maxtedford thanks for your contribution. It's very much appreciated.

Unfortunately my comments on the button borders diff are outdated. Did you read them, or should I copy the here into this discussion?

tvdeyen commented:
@maxtedford you can actually still see them [on the commit](https://github.com/AlchemyCMS/alchemy_cms/pull/1052/commits/12c573afacd588a64541f5b8e44e7ec41be19240)

maxtedford commented:
Okay @tvdeyen, I've added in some of the default button variables you mentioned as style overrides for the bourbon button mixin. I honestly didn't see a major change to the styles afterwards, but at the very least you'll have those variables to lean on when you inevitably migrate off of bourbon for your next major release. Sorry for all the back-and-forth here. Don't hesitate to let me know if you see any issues and I'd be happy to do any final tweaks. Thanks again!

tvdeyen commented:
Great. Thank you @maxtedford

We are almost finished, "just":

- Rebase on current `master`
- Address the issues HoundCI complains about (unused variable `e`) in `sass-support.rb`
- Please add the `sassc-rails` gem to the `Gemfile`, so we can start the dummy app.

Then I will make a final review and we are good to merge this. Great work! Thanks again for all the work.

maxtedford commented:
No problem @tvdeyen! Happy to help.

Here is the newly-rebased branch with the unnecessary `e` variable assignment removed and the `sassc-rails` gem added to the Gemfile.

Please let me know if you notice any last little hiccups during your final review, and thanks again for letting us contribute!

cc: @metaskills

metaskills commented:
WOW! So glad to be a part of this. Can I make one request. Can the `sass-support.rb` file follow Ruby convention and be `sass_support.rb` instead? So much Sass, I think we got file name convention mixed up :)

maxtedford commented:
Great catch, @metaskills! That is exactly what happened.

@tvdeyen, I'm gonna guess you'll be okay with me pushing up one last commit to address that naming convention. If not, I can revert it.

mamhoff commented:
One last nit: Would you mind rebasing out the back-and-forth commits? All the ones that first add something and then remove it can be amended IMHO. I'm actually quite happy with the small commit approach you used, but for example the renaming of `sass-support` to `sass_support` does not need to remain in our commit history. 

Sorry for bothering you more - this is excellent work, and the commit history is the last bit that's important here. The force push might help with the spec failure as well, which I believe is a spurious one.

maxtedford commented:
@mamhoff, no problem. Your point is completely valid. I took the liberty of squashing up some of the smaller back-and-forth commits, and reordering them to paint a little better picture of the full scope of the work. Hope that helps!

tvdeyen commented on app/assets/stylesheets/alchemy/base.scss:
```diff
@@ -142,31 +142,31 @@ kbd {
 
 .top-rounded-border {
-  @include border-top-radius($default-border-radius);
+  border-top-radius: $default-border-radius;
```
we should use [the Bourbon mixins](https://github.com/thoughtbot/bourbon/blob/v4-stable/app/assets/stylesheets/addons/_border-radius.scss) here, because they have a short hand for adding both `border-top-left-radius` and `border-top-right-radius` what we actually need here. Currently all "one sided" border radii went away from the GUI:

![alchemy cms - modal missing rounded borders](https://cloud.githubusercontent.com/assets/42868/16379270/d5eaabb8-3c70-11e6-8a63-17e660fe6030.jpg)

![alchemy cms - missing rounded borders](https://cloud.githubusercontent.com/assets/42868/16379273/db60df54-3c70-11e6-8c37-450145f5eb9b.jpg)

tvdeyen commented on app/assets/stylesheets/alchemy/base.scss:
```diff
@@ -142,31 +142,31 @@ kbd {
 
 .left-rounded-border {
-  @include border-left-radius($default-border-radius);
+  border-left-radius: $default-border-radius;
```
same here

tvdeyen commented on app/assets/stylesheets/alchemy/base.scss:
```diff
@@ -142,31 +142,31 @@ kbd {
 
 .right-rounded-border {
-  @include border-right-radius($default-border-radius);
+  border-right-radius: $default-border-radius;
```
etc

tvdeyen commented:
Nice, thanks. Still something left:

- [ ] Put "one-sided" border radii back (1)
- [ ] Add button styles to Tinymce buttons (2)
- [ ] Fix default button height (`29px`), by adjusting the padding to the former value (3)

### (1) Put "one-sided" border radii back

[See comment](https://github.com/AlchemyCMS/alchemy_cms/pull/1052/files#r68565398)

![alchemy cms - modal missing rounded borders](https://cloud.githubusercontent.com/assets/42868/16379429/d8c5d348-3c71-11e6-8ed3-cdc009e9368f.jpg)
![alchemy cms - missing rounded borders](https://cloud.githubusercontent.com/assets/42868/16379430/d8ebc5e4-3c71-11e6-8885-6bfa46c1a936.jpg)
![alchemy cms - rounded border linked button](https://cloud.githubusercontent.com/assets/42868/16379431/d8ed52a6-3c71-11e6-8ca8-246f62700bd7.jpg)

### (2) Add button styles to Tinymce buttons

Although I quite like the new flat look, we should at least have a hover state, what we don't have anymore. So, adding the default button styles back to tinymce buttons, should be the easiest for now.

![alchemy cms - tinymce buttons](https://cloud.githubusercontent.com/assets/42868/16379360/886580a6-3c71-11e6-8864-fe06cebf6b08.jpg)

### (3) Fix button height

The button height is incorrect, what would not be such a problem, if not the disabled button would jump now. It's value gets calculated via javascript when disabled. So, adjusting the padding should do the trick.

**wrong**
![alchemy cms - button height](https://cloud.githubusercontent.com/assets/42868/16379380/9c94f656-3c71-11e6-92c3-9b0c9e17b3de.jpg)

**correct**
![alchemy cms - correct button height](https://cloud.githubusercontent.com/assets/42868/16379464/ffe8969a-3c71-11e6-9003-0da9148a01a3.jpg)

If we address this, we are finally ready to merge. Thanks for cleaning the commit history and thanks @mamhoff for addressing this.

tvdeyen commented:
Oh, and could you please add `sassc-rails` into the global gem group? Otherwise Teatro.io (which runs in production env) does not know how to compile the assets, because there is no sass available, thanks.

maxtedford commented:
@tvdeyen, I've pushed up three new commits. Here's a brief rundown of the changes:

## **1. Put 'one-sided' border radii back** 

Great catch. I didn't even notice that. I incorporated the Bourbon mixin for each of the border-radii so that we have the desired border-radius effect in the current AlchemyCMS styles.

## **2. Add button styles to Tinymce buttons** 

I actually had some trouble recreating this on my end. Below is a .gif of what I see in my `spec/dummy` during development.

![alchemy_cms_tinymce_hover](https://cloud.githubusercontent.com/assets/9435299/16589705/47b0cf48-42a2-11e6-92b0-0ea3dbb7cb3b.gif)

From what I can see we've got hover effects and similar styles to the buttons and links elsewhere in the GUI. It's possible I'm doing something differently on my end that what you're doing, but I didn't actually make any changes to this since as far as I can tell it looks like it should be working.

## **3. Fix button height** 

This is all set as well. I overrode the Bourbon padding default from the `7px 18px` you saw in your console to the correct `0.5em 1.5em` you recommended. Let me know if anything looks off there.

## **4. Move sassc-rails gem into global gem group** 

I took care of this as well.

Let me know if you notice anything else that seems to be off, but hopefully we're all good at this point. Thanks again for the great software and for the opportunity to contribute. Looking forward to the next release!

cc: @metaskills

metaskills commented:
Love it!

tvdeyen commented on app/assets/stylesheets/tinymce/skins/alchemy/skin.min.css.scss:
```diff
@@ -616,12 +618,12 @@ body .mce-abs-layout-item,.mce-abs-end {
-  @include box-shadow(inset 0 1px 0 rgba(255,255,255,0.2),0 1px 2px rgba(0,0,0,0.05));
-  @include background(linear-gradient(color-stops($sb-base-color, darken($sb-base-color, 10%))));
+  box-shadow: inset 0 1px 0 rgba(255,255,255,0.2),0 1px 2px rgba(0,0,0,0.05);
+  @include background(linear-gradient(color-stops($white, darken($white, 10%))));
```
The syntax is actually without the `color-stops`:

```sass
@include background(linear-gradient($white, darken($white, 10%)));
```

see: http://bourbon.io/docs/#background

tvdeyen commented on app/assets/stylesheets/tinymce/skins/alchemy/skin.min.css.scss:
```diff
@@ -21,12 +22,13 @@
 .mce-widget button {
-  @include box-sizing(border-box);
+  box-sizing: border-box;
+  @include button($base-color: $white, $padding: none);
```
Please remove the `button` mix in from here. The actual styles are set on `.mce-btn` (see the comment below, because it wasn't working right now)

tvdeyen commented:
@maxtedford I found what causes the styling issues in tinymce, please see my inline comments.

tvdeyen commented on app/assets/stylesheets/tinymce/skins/alchemy/skin.min.css.scss:
```diff
@@ -638,10 +640,10 @@ body .mce-abs-layout-item,.mce-abs-end {
+      box-shadow: none;
       border: 0 none;
       background: none;
     }
```
We need to add an override for 

```sass
.mce-btn button:active { border: 0 none; }
```

Otherwise we will have duplicated borders when pressing the button:

![alchemy cms - all you can eat 2016-07-05 23-16-29](https://cloud.githubusercontent.com/assets/42868/16600321/d07624ae-4306-11e6-8202-44e411b73f03.jpg)

tvdeyen commented on app/assets/stylesheets/alchemy/_mixins.scss:
```diff
@@ -10,9 +10,16 @@
   border-color: $button-border-color;
-  margin: $default-form-field-margin;
+  font-weight: normal;
+  text-color: $text-color;
```
It's `color` not `text-color`

tvdeyen commented:
@maxtedford @metaskills https://github.com/customink/alchemy_cms/pull/1

metaskills commented:
Nice! Merged in and this PR has been updated.

maxtedford commented:
Wow, looks great! Thanks guys 🍻

tvdeyen commented:
Yes! Great work @maxtedford and @metaskills that's it!

Please once again rebase on current Alchemy master and force push this branch. 

Then we need to wait for 3.4 to land and we can merge this into the master and release it with 3.5

Very nice work. Can't wait to use sassc 🐎

tvdeyen commented:
Merged into master with 097d5334244c039d3928c30b7acb5b4c6954a567

Great work and thanks!

maxtedford commented:
Thanks @tvdeyen! I'm noticing that something may have gone wrong with the rebase...I'm not seeing the commits from your [bourbon-button-fixes](https://github.com/customink/alchemy_cms/pull/1) branch included, when I would have expected to see them at the end of the rebased commit history. I don't see them in my local reflog either, so perhaps they didn't fetch from the AlchemyCMS upstream correctly prior to the rebase. Maybe it's best to restore the branch and try again?

tvdeyen commented:
@maxtedford all commits landed in `master`  I used my local version of your branch to merge this. All fine. 

Thanks

maxtedford commented:
@tvdeyen, fantastic! Thanks again for the chance to contribute here. Really enjoyed it! 🎉
